### PR TITLE
[Query] Identify client using id @open sesame 9/2 14:47

### PR DESCRIPTION
**Pull request by gichan-jang**
currently, ip address of the client is only used to distinguish client which will receive buffers.
In this case, server cannot distinguish if more than one client with the samp ip is connected.
So instead of ip address, changed it to use the client's unique id.

Test pipeline
server:
```
gst-launch-1.0 tensor_query_serversrc ! other/tensors,num_tensors=1,dimensions=3:300:300:1,types=uint8,framerate=30/1 ! tensor_query_serversink
```
client 1:
```
gst-launch-1.0 v4l2src ! videoconvert ! videoscale !  video/x-raw,width=300,height=300,format=RGB,framerate=30/1 ! tensor_converter ! tensor_query_client ! tensor_decoder mode=direct_video ! videoconvert ! ximagesink
```
client2:
```
gst-launch-1.0 videotestsrc ! videoconvert ! videoscale !  video/x-raw,width=300,height=300,format=RGB,framerate=30/1 ! tensor_converter ! tensor_query_client ! tensor_decoder mode=direct_video ! videoconvert ! ximagesink
```

Signed-off-by: gichan <gichan2.jang@samsung.com>

**Self evaluation:**
1. Build test: [ *]Passed [ ]Failed [ ]Skipped
2. Run test: [ ]Passed [ ]Failed [*]Skipped

### CODE REVIEW

**Comment by jaeyun-jung** [comment]
@gichan-jang Plz check https://github.com/nnstreamer/api/pull/87. This will fix build error on Android.

File: gst/nnstreamer/tensor_query/tensor_query_common.c
Comment (wooksong) [inline gst/nnstreamer/tensor_query/tensor_query_common.c:256]: How about using a ```const``` variable instead of 37.

Comment (gichan-jang) [reply gst/nnstreamer/tensor_query/tensor_query_common.c]: Thanks. Updated.

File: gst/nnstreamer/tensor_query/tensor_query_common.c
Comment (dongju-chae) [inline gst/nnstreamer/tensor_query/tensor_query_common.c:257]: How about declaring this as macro?

File: gst/nnstreamer/tensor_query/tensor_query_common.c
Comment (dongju-chae) [inline gst/nnstreamer/tensor_query/tensor_query_common.c:260]: Well, it might be `uuid_size - 1`. Or, is there reason to include null chacter?

File: gst/nnstreamer/tensor_query/tensor_query_common.c
Comment (dongju-chae) [inline gst/nnstreamer/tensor_query/tensor_query_common.c:601]: No checking the return value?

File: gst/nnstreamer/tensor_query/tensor_query_common.c
Comment (gichan-jang) [reply gst/nnstreamer/tensor_query/tensor_query_common.c]: Changed as a macro.

File: gst/nnstreamer/tensor_query/tensor_query_common.c
Comment (gichan-jang) [reply gst/nnstreamer/tensor_query/tensor_query_common.c]: You're right.! Updated.

File: gst/nnstreamer/tensor_query/tensor_query_common.c
Comment (gichan-jang) [reply gst/nnstreamer/tensor_query/tensor_query_common.c]: Updated.

**Review by dongju-chae — APPROVED** [review]
LGTM

File: gst/nnstreamer/tensor_query/tensor_query_client.c
Comment (gichan-jang) [inline gst/nnstreamer/tensor_query/tensor_query_client.c:356]: uuid is supported from glib version 2.52.
If the version is too recent, I will generate the id as a manual.
@jaeyun-jung @again4you

**Review by jaeyun-jung — APPROVED** [review]

**Comment by gichan-jang** [comment]
DO NOT MERGE.
Client ID will be changed as uint32.

File: gst/nnstreamer/tensor_meta.h
Comment (myungjoo) [inline gst/nnstreamer/tensor_meta.h:28]: client_id as string?

Comment (gichan-jang) [reply gst/nnstreamer/tensor_meta.h]: Thanks. Change to uint32.

File: gst/nnstreamer/tensor_query/tensor_query_common.c
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_query/tensor_query_common.c:29]: plz fix it, CLIENT_ID_LEN 4

Comment (gichan-jang) [reply gst/nnstreamer/tensor_query/tensor_query_common.c]: Oh..Thanks. Updated.

**Comment by myungjoo** [comment]
The current mechanism generate ```client_id``` in client with random function and send it to server, which I don't think reliable enough.

I'd suggest to make client completely unaware of ```client_id```.

Proposal 1: if each connection is **persistent**.
1. client ---> server connected
2. server generates ```client_id``` for the newly created connection and saves ```client_id``` in the connection object.
3. server saves ```{ client_id,  connection_info }``` in the hash table.

Proposal 2: if each connection is **not persistent**. But I don't think it's this case.
1. client ---> server connected (new)
2. server generates ```client_id``` for the newly created connection and saves ```client_id``` in the connection object.
3. server --> client, send ```client_id```
4. server saves ```{ client_id,  connection_info }``` in the hash table.
5. client uses returned ```client_id```.

**Comment by myungjoo** [comment]
One thing you can consider as a principle for client/server systems:

- Server should (try to) be robust against malicious requests from clients.

With the original mechanism, if a client mimics another client by using ```client_id``` of another client, the service is broken.

**Comment by gichan-jang** [comment]
> The current mechanism generate `client_id` in client with random function and send it to server, which I don't think reliable enough.
> 
> I'd suggest to make client completely unaware of `client_id`.
> 
> Proposal 1: if each connection is **persistent**.
> 
> 1. client ---> server connected
> 2. server generates `client_id` for the newly created connection and saves `client_id` in the connection object.
> 3. server saves `{ client_id,  connection_info }` in the hash table.
> 
> Proposal 2: if each connection is **not persistent**. But I don't think it's this case.
> 
> 1. client ---> server connected (new)
> 2. server generates `client_id` for the newly created connection and saves `client_id` in the connection object.
> 3. server --> client, send `client_id`
> 4. server saves `{ client_id,  connection_info }` in the hash table.
> 5. client uses returned `client_id`.

Thanks,
There are two sockets between the server and the client. (client - serversrc, client-serversink)
Tensor query server use queue instead of hash table and don't share the data between server_src and server_sink.
How about this?
Supplement to proposal 2,
1. client ---> server_src connected (new)
2. server_src generates client_id for the newly created connection and saves client_id in the connection object.
3. server_src --> client, send client_id
4. server_src saves { client_id, connection_info } in the queue.
5. client ---> server_sink connected (new)
6. client ---> server_sink, send client_id and server_sink saves client_id in the connection object.
7. server_sink saves { client_id, connection_info } in the queue.
8. Connection complete.


> One thing you can consider as a principle for client/server systems:
> 
> * Server should (try to) be robust against malicious requests from clients.
> 
> With the original mechanism, if a client mimics another client by using `client_id` of another client, the service is broken.

Oh, I didn't think about it. Thanks.
I will also add a code to prevent this.

**Comment by myungjoo** [comment]
> > The current mechanism generate `client_id` in client with random function and send it to server, which I don't think reliable enough.
> > I'd suggest to make client completely unaware of `client_id`.
> > Proposal 1: if each connection is **persistent**.
> > 
> > 1. client ---> server connected
> > 2. server generates `client_id` for the newly created connection and saves `client_id` in the connection object.
> > 3. server saves `{ client_id,  connection_info }` in the hash table.
> > 
> > Proposal 2: if each connection is **not persistent**. But I don't think it's this case.
> > 
> > 1. client ---> server connected (new)
> > 2. server generates `client_id` for the newly created connection and saves `client_id` in the connection object.
> > 3. server --> client, send `client_id`
> > 4. server saves `{ client_id,  connection_info }` in the hash table.
> > 5. client uses returned `client_id`.
> 
> Thanks,
> There are two sockets between the server and the client. (client - serversrc, client-serversink)
> Tensor query server use queue instead of hash table and don't share the data between server_src and server_sink.
> How about this?
> Supplement to proposal 2,
> 
> 1. client ---> server_src connected (new)
> 2. server_src generates client_id for the newly created connection and saves client_id in the connection object.
> 3. server_src --> client, send client_id
> 4. server_src saves connection object in the queue.
> 5. client ---> server_sink connected (new)
> 6. client ---> server_sink, send client_id and server_sink saves client_id in the connection object.
> 7. server_sink saves connection object in the queue.
> 8. Connection complete.
> 

LGTM.

> > One thing you can consider as a principle for client/server systems:
> > 
> > * Server should (try to) be robust against malicious requests from clients.
> > 
> > With the original mechanism, if a client mimics another client by using `client_id` of another client, the service is broken.
> 
> Oh, I didn't think about it. Thanks.
> I will also add a code to prevent this.

As we have discussed some weeks ago (along with @jaeyun-jung and @again4you ), you will soon need a shared object (hash table) between server-src and server-sink.

**Comment by gichan-jang** [comment]
Updated to

**Review by myungjoo — APPROVED** [review]
LGTM. You may merge this when you are ready and comforatble.

**Review by jaeyun-jung — APPROVED** [review]

**Pull request merged**
